core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java

```diff
@@ -0,0 +1,340 @@
+import org.mule.runtime.core.internal.routing.outbound.EventBuilderConfigurer;
+import org.mule.runtime.core.internal.routing.outbound.EventBuilderConfigurerIterator;
+import org.mule.runtime.core.internal.routing.outbound.EventBuilderConfigurerList;
+import org.mule.runtime.core.internal.rx.FluxSinkRecorder;
+import org.mule.runtime.core.privileged.processor.chain.MessageProcessorChain;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import reactor.core.publisher.Flux;
+import reactor.util.context.Context;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.util.Optional.of;
+import static org.mule.runtime.api.functional.Either.left;
+import static org.mule.runtime.api.functional.Either.right;
+import static org.mule.runtime.api.metadata.DataType.fromObject;
+import static org.mule.runtime.core.api.event.CoreEvent.builder;
+import static org.mule.runtime.core.internal.routing.ExpressionSplittingStrategy.DEFAULT_SPLIT_EXPRESSION;
+import static org.mule.runtime.core.privileged.processor.MessageProcessors.applyWithChildContext;
+import static org.slf4j.LoggerFactory.getLogger;
+import static reactor.core.Exceptions.propagate;
+import static reactor.core.publisher.Mono.subscriberContext;
+import static reactor.util.context.Context.empty;
+
+class ForeachRouter {
+
+  private static final Logger LOGGER = getLogger(ForeachRouter.class);
+
+  private static final String MULE_FOREACH_CONTEXT_KEY = "mule.foreach.router.foreachContext";
+  static final String MAP_NOT_SUPPORTED_MESSAGE =
+      "Foreach does not support 'java.util.Map' with no collection expression. To iterate over Map entries use '#[dw::core::Objects::entrySet(payload)]'";
+  private final Map<String, ForeachContext> foreachContextResolver = new HashMap<>();
+  private final Foreach owner;
+
+  private Flux<CoreEvent> upstreamFlux;
+  private Flux<CoreEvent> innerFlux;
+  private Flux<CoreEvent> downstreamFlux;
+  private final FluxSinkRecorder<CoreEvent> innerRecorder = new FluxSinkRecorder<>();
+  private final FluxSinkRecorder<Either<Throwable, CoreEvent>> downstreamRecorder = new FluxSinkRecorder<>();
+  private final AtomicReference<Context> downstreamCtxReference = new AtomicReference(empty());
+
+  private final AtomicInteger inflightEvents = new AtomicInteger(0);
+  private final AtomicBoolean completeDeferred = new AtomicBoolean(false);
+
+  ForeachRouter(Foreach owner, Publisher<CoreEvent> publisher, String expression, int batchSize,
+                MessageProcessorChain nestedChain) {
+    this.owner = owner;
+
+    upstreamFlux = Flux.from(publisher)
+        .doOnNext(event -> {
+
+          if (expression.equals(DEFAULT_SPLIT_EXPRESSION)
```

Comment:
this is validation, not routing logic. This validation should be provided by the component.